\documentclass[11pt]{article}
\usepackage{sectsty}
\allsectionsfont{\color{blue}\fontfamily{lmss}\selectfont}
\usepackage{fontspec}
\setmainfont{XCharter}

\usepackage{listings}
\lstset{
basicstyle=\small\ttfamily,
tabsize=8,
columns=flexible,
breaklines=true,
frame=tb,
rulecolor=\color[rgb]{0.8,0.8,0.7},
backgroundcolor=\color[rgb]{1,1,0.91},
postbreak=\raisebox{0ex}[0ex][0ex]{\ensuremath{\color{red}\hookrightarrow\space}}
}
\usepackage{fontawesome}


\usepackage{mdframed}
\newmdenv[
  backgroundcolor=gray,
  fontcolor=white,
  nobreak=true,
]{terminalinput}



\usepackage{parskip}


    \usepackage[breakable]{tcolorbox}
    \usepackage{parskip} % Stop auto-indenting (to mimic markdown behaviour)

    \usepackage{iftex}
    \ifPDFTeX
    	\usepackage[T1]{fontenc}
    	\usepackage{mathpazo}
    \else
    	\usepackage{fontspec}
    \fi

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{format=nocaption,aboveskip=0pt,belowskip=0pt}

    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}
    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage{fancyvrb} % verbatim replacement that allows latex

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}



    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}

    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}


    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{index}





% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % For linebreaks inside Verbatim environment from package fancyvrb.
    \makeatletter
        \newbox\Wrappedcontinuationbox
        \newbox\Wrappedvisiblespacebox
        \newcommand*\Wrappedvisiblespace {\textcolor{red}{\textvisiblespace}}
        \newcommand*\Wrappedcontinuationsymbol {\textcolor{red}{\llap{\tiny$\m@th\hookrightarrow$}}}
        \newcommand*\Wrappedcontinuationindent {3ex }
        \newcommand*\Wrappedafterbreak {\kern\Wrappedcontinuationindent\copy\Wrappedcontinuationbox}
        % Take advantage of the already applied Pygments mark-up to insert
        % potential linebreaks for TeX processing.
        %        {, <, #, %, $, ' and ": go to next line.
        %        _, }, ^, &, >, - and ~: stay at end of broken line.
        % Use of \textquotesingle for straight quote.
        \newcommand*\Wrappedbreaksatspecials {%
            \def\PYGZus{\discretionary{\char`\_}{\Wrappedafterbreak}{\char`\_}}%
            \def\PYGZob{\discretionary{}{\Wrappedafterbreak\char`\{}{\char`\{}}%
            \def\PYGZcb{\discretionary{\char`\}}{\Wrappedafterbreak}{\char`\}}}%
            \def\PYGZca{\discretionary{\char`\^}{\Wrappedafterbreak}{\char`\^}}%
            \def\PYGZam{\discretionary{\char`\&}{\Wrappedafterbreak}{\char`\&}}%
            \def\PYGZlt{\discretionary{}{\Wrappedafterbreak\char`\<}{\char`\<}}%
            \def\PYGZgt{\discretionary{\char`\>}{\Wrappedafterbreak}{\char`\>}}%
            \def\PYGZsh{\discretionary{}{\Wrappedafterbreak\char`\#}{\char`\#}}%
            \def\PYGZpc{\discretionary{}{\Wrappedafterbreak\char`\%}{\char`\%}}%
            \def\PYGZdl{\discretionary{}{\Wrappedafterbreak\char`\$}{\char`\$}}%
            \def\PYGZhy{\discretionary{\char`\-}{\Wrappedafterbreak}{\char`\-}}%
            \def\PYGZsq{\discretionary{}{\Wrappedafterbreak\textquotesingle}{\textquotesingle}}%
            \def\PYGZdq{\discretionary{}{\Wrappedafterbreak\char`\"}{\char`\"}}%
            \def\PYGZti{\discretionary{\char`\~}{\Wrappedafterbreak}{\char`\~}}%
        }
        % Some characters . , ; ? ! / are not pygmentized.
        % This macro makes them "active" and they will insert potential linebreaks
        \newcommand*\Wrappedbreaksatpunct {%
            \lccode`\~`\.\lowercase{\def~}{\discretionary{\hbox{\char`\.}}{\Wrappedafterbreak}{\hbox{\char`\.}}}%
            \lccode`\~`\,\lowercase{\def~}{\discretionary{\hbox{\char`\,}}{\Wrappedafterbreak}{\hbox{\char`\,}}}%
            \lccode`\~`\;\lowercase{\def~}{\discretionary{\hbox{\char`\;}}{\Wrappedafterbreak}{\hbox{\char`\;}}}%
            \lccode`\~`\:\lowercase{\def~}{\discretionary{\hbox{\char`\:}}{\Wrappedafterbreak}{\hbox{\char`\:}}}%
            \lccode`\~`\?\lowercase{\def~}{\discretionary{\hbox{\char`\?}}{\Wrappedafterbreak}{\hbox{\char`\?}}}%
            \lccode`\~`\!\lowercase{\def~}{\discretionary{\hbox{\char`\!}}{\Wrappedafterbreak}{\hbox{\char`\!}}}%
            \lccode`\~`\/\lowercase{\def~}{\discretionary{\hbox{\char`\/}}{\Wrappedafterbreak}{\hbox{\char`\/}}}%
            \catcode`\.\active
            \catcode`\,\active
            \catcode`\;\active
            \catcode`\:\active
            \catcode`\?\active
            \catcode`\!\active
            \catcode`\/\active
            \lccode`\~`\~
        }
    \makeatother

    \let\OriginalVerbatim=\Verbatim
    \makeatletter
    \renewcommand{\Verbatim}[1][1]{%
        %\parskip\z@skip
        \sbox\Wrappedcontinuationbox {\Wrappedcontinuationsymbol}%
        \sbox\Wrappedvisiblespacebox {\FV@SetupFont\Wrappedvisiblespace}%
        \def\FancyVerbFormatLine ##1{\hsize\linewidth
            \vtop{\raggedright\hyphenpenalty\z@\exhyphenpenalty\z@
                \doublehyphendemerits\z@\finalhyphendemerits\z@
                \strut ##1\strut}%
        }%
        % If the linebreak is at a space, the latter will be displayed as visible
        % space at end of first line, and a continuation symbol starts next line.
        % Stretch/shrink are however usually zero for typewriter font.
        \def\FV@Space {%
            \nobreak\hskip\z@ plus\fontdimen3\font minus\fontdimen4\font
            \discretionary{\copy\Wrappedvisiblespacebox}{\Wrappedafterbreak}
            {\kern\fontdimen2\font}%
        }%

        % Allow breaks at special characters using \PYG... macros.
        \Wrappedbreaksatspecials
        % Breaks at punctuation characters . , ; ? ! and / need catcode=\active
        \OriginalVerbatim[#1,codes*=\Wrappedbreaksatpunct]%
    }
    \makeatother

    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}

    % prompt
    \makeatletter
    \newcommand{\boxspacing}{\kern\kvtcb@left@rule\kern\kvtcb@boxsep}
    \makeatother
    \newcommand{\prompt}[4]{
        {
        \ttfamily\llap{{\color{blue}\LARGE\faKeyboardO\hspace{3pt}#4}}\vspace{-\baselineskip}
	}
    }



    % Prevent overflowing lines due to hard-to-break entities
    \sloppy
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults

    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}



\renewcommand{\PY}[2]{{#2}}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{\color{gray}\sf\small\rightmark}
\lhead{\nouppercase{\color{gray}\sf\small\leftmark}}
\cfoot{\color{gray}\sf\thepage}
\renewcommand{\footrulewidth}{1pt}
\begin{document}





    \hypertarget{phylogeny-and-mapping}{%
\section{Phylogeny and Mapping}\label{phylogeny-and-mapping}}

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

Phylogenetics is essentially about similarity, and looking at patterns
of similarity between taxa to infer their relationships. It has
important applications in many fields of genome biology. For example,
when annotating a gene in a new genome it is useful for identifying
previously-annotated genes in other genomes that share a common
ancestry. It is also becoming increasingly common to use phylogeny to
trace the evolution and spread of bacterial diseases, and even as an
epidemiological tool to help identify disease outbreaks in a clinical
setting. Further analysis of genome sequences to examine recombination,
molecular adaptation and the evolution of gene function, all benefit
from phylogeny.

\hypertarget{learning-outcomes}{%
\subsection{Learning outcomes}\label{learning-outcomes}}

On completion of the tutorial, you can expect to be able to:

\begin{itemize}
\tightlist
\item
  Identify different approaches to constructing phylogeny from whole
  genome sequence data
\item
  Map sequence data to a reference genome and identify variants
\item
  Use SNP data to construct a phylogenetic tree
\item
  Identify and remove recombination with Gubbins
\item
  Visualise phylogenies in the context of the sample metadata
\end{itemize}

\hypertarget{tutorial-sections}{%
\subsection{Tutorial sections}\label{tutorial-sections}}

This tutorial comprises the following sections:\\
1. \href{intro_to_phylogeny.ipynb}{Introduction to phylogenetics}\\
2. \href{gene.ipynb}{Phylogeny from gene sequences}\\
3. \href{snp_phylogeny.ipynb}{Phylogeny from whole genome sequence
data}\\
4. \href{metadata.ipynb}{Phylogeny and metadata}

\hypertarget{authors-and-license}{%
\subsection{Authors and License}\label{authors-and-license}}

This tutorial was written by
\href{https://github.com/jacquikeane}{Jacqui Keane}.

Some of the material has been adapted from the Wellcome Connecting
Science Courses
\href{https://github.com/WCSCourses/AMR-Asia-23}{AMR-ASIA23},
\href{https://github.com/WCSCourses/GenEpiLAC2023}{GenEpiLAC2023} and
\href{https://github.com/WCSCourses/WWPG_2021}{WWPG21}.

The content is licensed under a
\href{https://creativecommons.org/licenses/by/4.0/}{Creative Commons
Attribution 4.0 International License (CC-By 4.0)}.

\hypertarget{running-the-commands-in-this-tutorial}{%
\subsection{Running the commands in this
tutorial}\label{running-the-commands-in-this-tutorial}}

You can follow this tutorial by typing all the commands you see in a
terminal window on your computer. Remember, the terminal window is
similar to the ``Command Prompt'' window on MS Windows systems, which
allows the user to type DOS commands to manage files.

To get started, open a terminal window and type the command below
followed by the \texttt{Enter} key:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} \PYZti{}/course\PYZus{}data/snp\PYZhy{}phylogeny/data
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{prerequisites}{%
\subsection{Prerequisites}\label{prerequisites}}

This tutorial assumes that you have the following software and their
dependencies installed on your computer. The software used in this
tutorial may be updated from time to time so, we have also given you the
version which was used when writing this tutorial.

\begin{longtable}[]{@{}ccc@{}}
\toprule
Package name & Link for download/installation instructions &
Version\tabularnewline
\midrule
\endhead
seaview & https://doua.prabi.fr/software/seaview & 5.0.5\tabularnewline
fastqc & https://www.bioinformatics.babraham.ac.uk/projects/fastqc &
0.12.1\tabularnewline
fastp & https://github.com/OpenGene/fastp & 0.23.4\tabularnewline
bwa & https://github.com/lh3/bwa & 0.7.17\tabularnewline
samtools & https://github.com/samtools/samtools & 1.17\tabularnewline
bcftools & https://github.com/samtools/bcftools & 1.17\tabularnewline
snp-sites & https://github.com/sanger-pathogens/snp-sites &
2.5.1\tabularnewline
gubbins & https://github.com/nickjcroucher/gubbins &
3.3.0\tabularnewline
iqtree & http://www.iqtree.org/ & 2.2.3\tabularnewline
FigTree & http://tree.bio.ed.ac.uk/software/figtree/ &
1.4.4\tabularnewline
Microreact & https://microreact.org & 240\tabularnewline
\bottomrule
\end{longtable}

    The easiest way to install the required software is using
\texttt{conda}, a software package manager. These software have already
been installed on the computer for you. To activate them type:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
conda activate snp\PYZhy{}phylogeny
\end{Verbatim}
\end{tcolorbox}

    After the software is activated type the following commands:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
seaview \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
fastqc \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
fastp \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bwa
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
snp\PYZhy{}sites \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
run\PYZus{}gubbins.py \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
iqtree \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
figtree \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    This should return the help message for these tools or launch the GUI
software in the background.

To get started with the tutorial, go to the first section:
\href{intro_to_phylogeny.ipynb}{Introduction to phylogeny}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{introduction-to-phylogeny}{%
\section{Introduction to Phylogeny}\label{introduction-to-phylogeny}}

Before we describe how to construct a phylogenetic tree, we will
introduce some important concepts related to phylogenies.

    \hypertarget{what-is-a-phylogentic-tree}{%
\subsection{What is a phylogentic
tree?}\label{what-is-a-phylogentic-tree}}

A phylogeny, also known as a phylogenetic tree, depicts estimated
evolutionary relationships between \textbf{taxa} - these can be species,
strains or even genes. It consists of \textit{nodes} and \textit{branches}.


\begin{center}
\includegraphics[A phylogenetic tree]{img/basic-tree.png}
\end{center}


\hypertarget{nodes}{%
\subsubsection{Nodes}\label{nodes}}

There are two types of \textit{nodes} in a phylogenetic tree; external
nodes, also called `tips' or `leaves' of the tree (i.e.~taxa), and
internal nodes which denote their hypothetical ancestors.

\hypertarget{branches}{%
\subsubsection{Branches}\label{branches}}

The \textit{branches} in a tree are shown as lines connecting the nodes.
They show the path of transmission of genetic information from one
generation to the next. Branch lengths indicate genetic change i.e.~the
longer the branch, the more genetic change (or divergence) has occurred.

\hypertarget{root}{%
\subsubsection{Root}\label{root}}

A phylogenetic tree can be unrooted or rooted as shown below.


\begin{center}
\includegraphics[A phylogenetic tree]{img/unrooted_vs_rooted.png}
\end{center}


The \textit{root} of a phylogenetic tree is the most recent common
ancestor of all of the taxa in the tree. It is therefore the oldest part
of the tree and tells us the direction of evolution, with the flow of
genetic information moving from the root, towards the tips with each
successive generation. Deciding upon an appropriate root position is
critical for phylogenetic interpretation because the root tells us the
direction of evolution and so affects statements that we make about
patterns of relatedness.

There are two main approaches that can be used to root a tree, midpoint
rooting and outgroup rooting. We will describe these approaches in more
detail later in the tutorial.

\hypertarget{topology}{%
\subsubsection{Topology}\label{topology}}

The \textit{topology} is the branching structure of the tree and the same
topology can be drawn in lots of different ways. It is of particular
biological significance because it indicates patterns of relatedness
among taxa, meaning that trees with the same topology and root have the
same biological interpretation.

    \hypertarget{microbial-phylogenetic-trees}{%
\subsection{Microbial phylogenetic
trees}\label{microbial-phylogenetic-trees}}

In the context of infectious diseases epidemiology, phylogenetic trees
are commonly used to define evolutionary relationships between strains
of the same bacterial species. This is possible because bacteria
reproduce clonally. During clonal reproduction, bacterial progenitor
cells replicate their DNA at high fidelity. Despite this, random errors
in DNA replication may still occur, resulting in a clonal progeny that
will inherit these genetic replication `errors' (i.e.~mutations) in
their DNA and may not be strictly identical to their progenitor cells.
Bacterial strains that have recently originated from the same progenitor
cell are thus expected to share identical genomes or have diverged at
most by only a few genetic differences (mutations). The number and
pattern of shared mutations between bacterial strains can be used to
reconstruct their genealogical and evolutionary relationships.

On a phylogenetic tree, isolated bacterial strains are depicted on the
\textit{leaves} of the tree, whereas the internal nodes of the tree denote
their hypothetical ancestors. Groups of bacterial strains (taxa) that
share the same common ancestor form a \textit{monophyletic} group (also
known as \textit{clade}). A group of strains that descends from a common
ancestor, but does not include all descendants, is called
\textit{paraphyletic}.

    \hypertarget{how-are-phylogenetic-trees-reconstructed}{%
\subsection{How are phylogenetic trees
reconstructed?}\label{how-are-phylogenetic-trees-reconstructed}}

Today almost all phylogenetic trees are inferred from molecular sequence
data, most often from DNA sequences. This is because DNA is an inherited
material; it can easily, reliably and inexpensively be extracted and
sequenced; and DNA sequences are highly specific to bacterial species
and strains.

The goal is to generate a multiple sequence alignment of the DNA
sequences of the bacterial species or strains and use this to
re-construct the phylogenetic tree. Polymorphic sites (that is,
nucleotide positions that are variable across multiple strains) in
multiple alignments are used to infer evolutionary relationships,
whereas monomorphic sites (nucleotide positions with the same DNA base)
are generally ignored. The figure below shows an example of a simple
multiple alignment of eight sites from four strains, which include
monomorphic (squared) and polymorphic sites. Genetic changes in the
phylogenetic tree are showed as coloured vertical rectangles on the
branch where they originated. The identification of genetic changes
(alleles) that are unique and common to multiple taxa (strains) are used
to group them into monophyletic groups (clades) in a hierarchical manner
with the goal of constructing the most plausible genealogical
relationships between strains and clades.


\begin{center}
\includegraphics[Overview of phylogeny construction]{img/phylogeny.png}
\end{center}


    There are many computational methods for reconstructing a tree from
sequence data, e.g.~Parsimony, Maximum Likelihood. Most methods require
a model of evolution (substitution model) to provide information on how
the sequences evolved. We won't expand on this here as it is beyond the
scope of this tutorial.

    \hypertarget{data-selection-for-phylogenies}{%
\subsection{Data selection for
phylogenies}\label{data-selection-for-phylogenies}}

The type of sequence data that is used to construct a phylogenetic tree
can vary. A tree can be constructed using sequence data from:

\begin{itemize}
\tightlist
\item
  a single gene
\item
  Multi-locus sequence typing (MLST) data: a set of housekeeping genes
  for a species
\item
  Core genome MLST (cgMLST) data: an expanded MLST scheme using all
  genes shared by all strains within a species
\item
  Whole genome MLST (wgMLST) data: an expanded cgMLST scheme to include
  a selection of accessory genes
\item
  entire genome reconstructed from the whole genome sequence data
\end{itemize}

The sequences to include and the type of data that you choose when
constructing your phylogenetic tree will depend on your objective.

    \hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

Look at the following tree and answer the questions that follow:


\begin{center}
\includegraphics[Overview of phylogeny construction]{img/test-tree.png}
\end{center}


\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many nodes are in the tree?
\item
  What internal node corresponds to the most recent common ancestor of
  samples 8 and 10
\item
  Which term best describes a group that share the same common ancestor?
\end{enumerate}

\begin{enumerate}
\def\labelenumi{\alph{enumi}.}
\tightlist
\item
  Polyphyletic
\item
  Monophyletic
\item
  Paraphyletic
\end{enumerate}

    Now let's construct our first phylogeny: \href{gene.ipynb}{Phylogeny
from gene sequences}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{phylogeny-from-gene-sequences}{%
\section{Phylogeny from gene
sequences}\label{phylogeny-from-gene-sequences}}

For our first phylogeny we will make a tree using the sequences of the
\textit{ompA} gene from 16 strains of \textit{C. trachomatis}. First, let's
check that you're in the right place. Type the command below in the
terminal window followed by the Enter key:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{pwd}
\end{Verbatim}
\end{tcolorbox}

    It should display something like:

\texttt{/home/manager/course\_data/snp-phlogeny/data}

The gene sequences for the 16 strains are found in a multifasta file
called \texttt{ompA.fa} in the \texttt{gene} directory. These nucleotide
sequences have been constructed from sequence data of the 16 strains. We
won't cover how the genes were constructed here but will learn more
about this in a later tutorial.

Take a look at the file containing the gene sequences:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} gene
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat ompA.fa
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{introducing-the-dataset}{%
\subsection{Introducing the dataset}\label{introducing-the-dataset}}

The dataset used in this section has in part been derived from the
following paper:

\begin{quote}
\textbf{The Swedish new variant of Chlamydia trachomatis: genome
sequence, morphology, cell tropism and phenotypic characterization}\\
Unemo M, Seth-Smith HMB, Cutcliffe LT, Skilton RJ, Barlow D, Goulding D,
Persson K, Harris SR, Kelly A, Bjartling C, Fredlund H, Olcén P, Thomson
NR, Clarke IN. \textit{Microbiology 2010. doi: 10.1099/mic.0.036830-0}
PMID:
\href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3541825/}{20093289}
\end{quote}

\textit{Chlamydia trachomatis} is one of the most prevalent human
pathogens in the world, causing a variety of infections. It is the
leading cause of sexually transmitted infections (STIs), with an
estimated 131 million new cases each year. Additionally, it is also the
leading cause of preventable infectious blindness with tens of millions
of people thought to have active disease.

Historically, the most commonly used tool for typing \textit{C.
trachomatis} isolates was serotyping using the MOMP (major outer
membrane protein), which is encoded by the \textit{ompA} gene. There are
two biovars of C. trachomatis:

\begin{itemize}
\tightlist
\item
  the \textit{trachoma} biovar includes ocular and urogenital strains,
  which cause the majority of trachoma and STIs, and are characterised
  by localised infections of the epithelial surface of the conjunctiva
  or genital mucosa;
\item
  the \textit{lymphogranuloma venereum (LGV)} biovar includes strains
  which are distinguished by their ability to spread systemically
  thorough the lymphatic system, causing genital ulceration and bubonic
  disease.
\end{itemize}

Based on MOMP serotyping, \textit{C. trachomatis} has been subdivided into
between 15 and 19 serotypes: the \textit{trachoma} biovar includes ocular
serotypes A to C and urogenital serotypes D to K, while the \textit{LGV}
biovar includes serotypes L1, L2 (including L2a, b and c) and L3.

In 2006, a new variant of \textit{C. trachomatis}, known as NV was
reported in Sweden and triggered a European health alert. During this
time it became the dominant strain circulating in some European
countries and began to spread world wide. The reason for this was that
it evaded detection by the widely used PCR-based diagnostic test. During
this exercise we will place this New Variant of \textit{C. trachomatis}
(NV) in the context of several other \textit{C.trachomatis} strains based
on the ompA gene.

    \hypertarget{viewing-the-alignment-in-seaview}{%
\subsection{Viewing the alignment in
Seaview}\label{viewing-the-alignment-in-seaview}}

To view our gene sequences and produce a phylogeny we will use a program
called \texttt{Seaview}. \texttt{Seaview} is a graphical user interface
(GUI) that combines a number of the most popular alignment and phylogeny
programs. Launch \texttt{Seaview}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
seaview \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    Now load the file \texttt{ompA.fa} by selecting `File' and `Open' from
the main menu.


\begin{center}
\includegraphics[Seaview]{img/seaview_1.png}
\end{center}



\begin{center}
\includegraphics[ompA alignment in Seaview]{img/seaview_2.png}
\end{center}


    \hypertarget{multiple-sequence-alignment}{%
\subsection{Multiple sequence
alignment}\label{multiple-sequence-alignment}}

Homology among DNA, RNA, or proteins is typically inferred from their
nucleotide or amino acid sequence similarity. Significant similarity is
strong evidence that two sequences are related by evolutionary changes
from a common ancestral sequence. Alignments of multiple sequences are
used to indicate which regions of each sequence are homologous.

Therefore before we perform any phylogenetic analysis, we must make sure
that the columns in our data represent homologous bases. With gene or
protein sequence data, this usually means aligning the nucleotide or
amino acid sequences using a multiple alignment program. Length
differences of the sequences complicate multiple sequence alignment
because these require the insertion of gaps into an alignment to ensure
that homologous sites remain aligned. When possible, alignments should
be checked by eye.

\texttt{Seaview} allows alignment using two programs, \texttt{clustal}
and \texttt{muscle}. Generally \texttt{muscle} is faster, providing
protein alignments that are of similar quality to \texttt{clustal}. It
is usually better to align genes after translating them into amino
acids, so we will do that here.

    On the `Props' menu tick `View as proteins'.


\begin{center}
\includegraphics[View as proteins in Seaview]{img/seaview_3.png}
\end{center}


    To run the alignment, select `Align' then `Align all' from the main
menu.


\begin{center}
\includegraphics[Aign all proteins in Seaview]{img/seaview_4.png}
\end{center}


When the alignment process is complete, \texttt{Seaview} will have
inserted gaps into the sequences so that homologous sites (or at least
homologous according to the alignment program) are lined up in columns.
Look at the alignment, it should now be clearer how the sequences differ
from one another.


\begin{center}
\includegraphics[Protein alignment in Seaview]{img/seaview_5.png}
\end{center}


    Now turn off protein view and you will see that the nucleotides are also
now aligned.


\begin{center}
\includegraphics[Nucleotide alignment in Seaview]{img/seaview_6.png}
\end{center}


    \hypertarget{phylogeny-estimation-using-phyml}{%
\subsection{Phylogeny estimation using
PhyML}\label{phylogeny-estimation-using-phyml}}

To estimate the phylogeny, we will use a program called \texttt{PhyML},
which is included in \texttt{Seaview}. \texttt{PhyML} uses a maximum
likelihood (ML) method to estimate the tree and includes a number of
nucleotide substitution models ranging from the very simple (and could
be unrealistic) to more complex ones. Evolutionary (or substitution)
models are statistical models that describe the substitution and
divergence of sequences over time.

To create a tree in \texttt{Seaview}, select `Trees' and `PhyML' from
the main menu.


\begin{center}
\includegraphics[Creating a treee with PhyML]{img/phyml.png}
\end{center}


    Choose the \textit{GTR} (General Time Reversible) model for substitution
rates, set all other parameters as shown below and click \texttt{Run}.


\begin{center}
\includegraphics[Tree Settings]{img/tree_settings.png}
\end{center}


    Once the run has finished, click `OK' and you should see a phylogenetic
tree as shown below. The tree created by \texttt{PhyML} includes the
topology of tree (i.e., the relationships between sequences) and the
branch lengths (i.e., the amount of change occurring in each lineage).
Therefore, the tree is drawn as a \textit{phylogram}, in which the length
of branches is proportional to the amount of evolutionary change.


\begin{center}
\includegraphics[Phylogenetic tree of ompA gene with detailed description]{img/tree_with_descriptions.png}
\end{center}


    \hypertarget{phylogeny-estimation-with-bootstrapping}{%
\subsection{Phylogeny estimation with
bootstrapping}\label{phylogeny-estimation-with-bootstrapping}}

Bootstrapping is a statistical technique to assess the confidence level
around each phylogenetic node. The bootstrapping values indicates how
many times out of 100, the same branch was observed when repeating the
phylogenetic reconstruction on a re-sampled set of the data. Robust
relationship should be repeatable, and subsequently observed in a large
proportion of randomised data. Therefore, if you get 100 out of 100
times for a particular node, you can be more confident that the observed
branch is not due to chance, but likely to be real.

To estimate a bootstrapped phylogeny for the \textit{ompA} data, start by
creating a new phylogeny as before and then click on `Bootstrap' in the
`Branch Support' box, and enter `10' in the replicates box. The
processing of the 10 replicates may take a few minutes, so you could
move on while this is running. Ideally you would run more replicates
(100 or 1000), but to save time we will only run 10 replicates here.


\begin{center}
\includegraphics[Bootstrapping option]{img/bootstrap.png}
\end{center}


Once the search is complete, you can show the bootstrap values on the
tree by ticking the \textbf{Br support} box.


\begin{center}
\includegraphics[Phylogenetic tree with bootstraps]{img/tree_with_bootstraps.png}
\end{center}


Each node in the tree now has an associated value out of 100, its
bootstrap. Can you identify any nodes that are not robust? Unfortunately
there is no generally accepted threshold for significant bootstrap
robustness, so you must use your judgement.

\textbf{WARNING!:} bootstrap proportions are measures of robustness, or
repeatability. A high bootstrap value indicates that a given node tends
to occur in every analysis. This does not guarantee that the node is
correct. For example, if the substitution model is inaccurate, it could
produce the wrong answer in every estimation.

    \hypertarget{phylogeny-visualisation}{%
\subsection{Phylogeny visualisation}\label{phylogeny-visualisation}}

In addition to \texttt{Seaview} there are many other tools that can be
used to visualise phylogenetic trees including \texttt{FigTree},
\texttt{Microreact} and \texttt{iToL}. As a short exercise, let's use
\texttt{FigTree} to visualise our tree. \texttt{FigTree} is more
versatile than the tree viewer in \texttt{Seaview}, allowing you to
colour branches and taxa, redraw the tree in a number of ways and output
the results in a large number of graphics formats including \texttt{eps}
and \texttt{pdf}. It is particularly useful for preparing figures for
manuscripts.

First, export the tree produced by PhyML by choosing the `File' and Save
unrooted tree' from the main menu. The file will be saved in newick
format. A newick file is a standard text-based format for representing
trees in computer-readable form using (nested) parentheses and commas.

Open FigTree by typing `figtree' in the terminal.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
figtree \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    Now open the exported newick tree file using \texttt{FigTree}.


\begin{center}
\includegraphics[Open newick file with FigTree]{img/figtree.png}
\end{center}


If you have time explore the different functionality available in
\texttt{FigTree}.


\begin{center}
\includegraphics[Open newick file with FigTree]{img/tree_in_figtree.png}
\end{center}


    \hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  From the trees that you have produced, which MOMP type would you
  suggest the new variant (NV) strain belongs to?
\item
  Do the ompA trees agree with the separation of \textit{C. trachomatis}
  into trachoma (serotypes A to K) and LGV (L serotypes) biovars?
\end{enumerate}

    Now move on to the next section: \href{snp_phylogeny.ipynb}{Constructing
phylogeny from whole genome sequence data}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{phylogeny-from-whole-genome-sequence-data}{%
\section{Phylogeny from whole genome sequence
data}\label{phylogeny-from-whole-genome-sequence-data}}

    When we sequence a population, we aim to capture the variation (SNPs,
indels, gene gain and loss etc.) in the samples and use it to infer the
relationships between the samples. Two of the main approaches to
capturing this variation and reconstructing the bacterial genomes are:

\begin{itemize}
\tightlist
\item
  De novo genome assembly and annotation
\item
  Mapping and variant calling against a reference genome
\end{itemize}

Each approach has it's benefits and limitations. We will focus on
mapping and variant calling in this tutorial. For mapping and variant
calling, whether we are dealing with different bacterial isolates, with
viral populations in a patient, or even with genomes of different human
individuals, the principles are essentially the same. Instead of
assembling the newly generated sequence reads de novo to produce a new
genome sequence, it is easier and much faster to align or map the
sequence reads to a reference genome. We can then readily identify SNPs
and indels that distinguish closely related populations or individual
organisms and may thus learn about genetic differences that may cause
drug resistance or increased virulence in pathogens, or changed
susceptibility to disease in humans. One important prerequisite for the
mapping of sequence data to work is that the reference and the
re-sequenced subject have the same genome architecture.

In this exercise, we will use sequence data from \textit{Salmonella
enterica serovar Typhi} samples to demonstrate the mapping and variant
calling approach. Importantly, although the data is based on real
sequence data, it has been edited to make it run more efficiently for
the purpose of this tutorial.

Navigate to the directory that contains the sequence data:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} \PYZti{}/course\PYZus{}data/snp\PYZhy{}phylogeny/data/typhi
\end{Verbatim}
\end{tcolorbox}

    Take a look at the directory containing the sequence data for the
samples:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls fastq
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{introducing-the-tutorial-dataset}{%
\subsection{Introducing the tutorial
dataset}\label{introducing-the-tutorial-dataset}}

We will use data adapted from the following paper:

\begin{quote}
\textbf{A genomic snapshot of Salmonella enterica serovar Typhi in
Colombia}\\
Guevara, Paula Diaz, et al.\\
\textit{PLoS Neglected Tropical Diseases2021. doi:
10.1371/journal.pntd.0009755}\\
PMID:
\href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8478212/}{34529660}
\end{quote}

\textit{Salmonella enterica serovar Typhi} (\textit{S. Typhi}) is the
causative agent of typhoid fever, with between 9--13 million cases and
116,800 associated deaths annually. Typhoid fever is still a public
health problem in many countries, including in Latin America, which has
a modelled incidence of up to 169 (32--642) cases per 100,000
person-years. Several international studies have aimed to fill data gaps
regarding the global distribution and genetic landscape of typhoid;
however, in spite of these efforts Latin America is still
underrepresented. This study provided the first enhanced insights into
the molecular epidemiology of S. Typhi in Colombia, using whole genome
sequencing data to investigate the population structure in Colombia and
identify predominant circulating genotypes.

    \hypertarget{overview-of-mapping-and-variant-calling-approach}{%
\subsection{Overview of mapping and variant calling
approach}\label{overview-of-mapping-and-variant-calling-approach}}

The diagram below illustrates the steps involved when mapping and
calling variants for a set of bacterial samples.


\begin{center}
\includegraphics[Approach]{img/snp-phylogeny-approach.png}
\end{center}


The first step once you have obtained your sequence data (FASTQ) is to
QC the data. After QC, the sequence data is matched or aligned to a
reference genome (FASTA) in a process called read mapping to produce a
set of read aligments (SAM/BAM). These read alignments are inspected to
identify differences between the aligned reads and the reference genome.
This process is called variant calling and produces VCF files. In fact
during this process we capture information about every position in the
genome (variant and non-variant sites) in the VCFs. Each site in the VCF
has a set of quality filters applied and any sites identified as low
quaility (e.g.~less than 4 reads aligned at that position) are marked as
low quality in the VCF to produce a filtered VCF. We use this filtered
VCF file in a process called consensus caling to reconstruct a consensus
\textit{pseudosequence} or \textit{pseudogenome} for our sample (FASTA). In
the \textit{pseudogenome}, any sites marked as low quality will be
represented as an N in the reconstructed sequence. These pseudogenomes
(multi-FASTA) are then aligned and variation identified and used to
reconstruct a phylogeny of our samples.

    \hypertarget{exercise}{%
\subsection{Exercise}\label{exercise}}

Now let's analyse some data!

\hypertarget{prepare-the-data}{%
\subsubsection{Prepare the data}\label{prepare-the-data}}

First take a look at the sequence data provided.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls fastq/
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many samples have been sequenced?
\item
  How many fastq files are there?
\end{enumerate}

    We will use the chromosome sequence of \textit{Salmonella typhi CT18} as
the reference genome. This has already been downloaded from RefSeq. Take
a look at the reference genome:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ref/
\end{Verbatim}
\end{tcolorbox}

    Check the size of the reference file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
assembly\PYZhy{}stats ref/Styphi\PYZus{}CT18.fa
\end{Verbatim}
\end{tcolorbox}

    Now use bwa to index the reference genome. This creates a lookup table
that bwa uses when matching the sequence reads against the reference
genome.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bwa index ref/Styphi\PYZus{}CT18.fa
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  How many sequences in the reference fasta file?
\item
  What are the names of the sequences in the reference fasta file?
\item
  What is the size of the reference?
\item
  What additional files did the indexing step produce?
\end{enumerate}

    \hypertarget{qc-the-sequence-data}{%
\subsubsection{QC the sequence data}\label{qc-the-sequence-data}}

An important first step in any analysis is QC of the data. We will the
FastQC software to QC the data. First create a directory for the qc
results:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} fastq
mkdir qc\PYZus{}results
\end{Verbatim}
\end{tcolorbox}

    Run FastQC on the all the fastq files and store the results in the
directory \texttt{qc\_results}:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
fastqc \PYZhy{}o qc\PYZus{}results *.fastq.gz
\end{Verbatim}
\end{tcolorbox}

    This will create one html report for each of the fastq files. Take a
look:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls qc\PYZus{}results/
\end{Verbatim}
\end{tcolorbox}

    If we have lots of samples it will be difficult to manually inspect each
file. Therefore we will use multiQC to collate all the QC reports into
one file.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
multiqc qc\PYZus{}results
\end{Verbatim}
\end{tcolorbox}

    Open the collated report \texttt{multiqc\_report.html} in firefox.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
firefox multiqc\PYZus{}report.html \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    \begin{figure}
\centering
\includegraphics{img/multiqc.png}
\caption{MultiQC results}
\end{figure}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  What is the median read length for sample ERR5243693?
\item
  Which sample has the largest yield (most sequence data)?
\end{enumerate}

    \hypertarget{trim-the-reads-to-remove-low-quality-and-adapter-sequence}{%
\subsubsection{Trim the reads to remove low quality and adapter
sequence}\label{trim-the-reads-to-remove-low-quality-and-adapter-sequence}}

If your sequence reads have a high level of adapter contamination and/or
have low quality bases at the end of the reads you can trim the reads to
remove these sequences. There are several software that can be used to
do this including trimmomatic and fastp.

Use fastp to trim the reads for sample ERR5243693.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
fastp \PY{l+s+se}{\PYZbs{}}
   \PYZhy{}\PYZhy{}in1 ERR5243693\PYZus{}1.fastq.gz \PYZhy{}\PYZhy{}in2 ERR5243693\PYZus{}2.fastq.gz \PY{l+s+se}{\PYZbs{}}
   \PYZhy{}\PYZhy{}out1 ERR5243693\PYZus{}1.trim.fastq.gz \PY{l+s+se}{\PYZbs{}}
   \PYZhy{}\PYZhy{}out2 ERR5243693\PYZus{}2.trim.fastq.gz \PY{l+s+se}{\PYZbs{}}
   \PYZhy{}\PYZhy{}json ERR5243693.fastp.json \PYZhy{}\PYZhy{}html ERR5243693.fastp.html \PY{l+s+se}{\PYZbs{}}
   \PYZhy{}\PYZhy{}detect\PYZus{}adapter\PYZus{}for\PYZus{}pe \PYZhy{}\PYZhy{}cut\PYZus{}mean\PYZus{}quality \PY{l+m}{20} \PY{l+s+se}{\PYZbs{}}
   \PYZhy{}\PYZhy{}thread \PY{l+m}{2}
\end{Verbatim}
\end{tcolorbox}

    Now repeat for the other samples and look at the output from fastp:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head \PYZhy{}20 ERR5243693.fastp.json
head \PYZhy{}20 ERR5243695.fastp.json
head \PYZhy{}20 ERR5243699.fastp.json
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How much data (bp/base pairs) was lost due to trimming and adapter
  removal?
\end{enumerate}

    \hypertarget{map-the-data-to-a-reference-genome}{%
\subsubsection{Map the data to a reference
genome}\label{map-the-data-to-a-reference-genome}}

    When performing any data analysis it is good practice to arrange your
data in a logical way rather than putting all files in a single
directory. Let's create a directory to store the results of mapping the
reads to the reference genome:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
mkdir ../samtools
\PY{n+nb}{cd} ../samtools
\end{Verbatim}
\end{tcolorbox}

    Use \texttt{bwa} to map the reads for sample ERR5243693 to the reference
genome.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bwa mem \PYZhy{}t \PY{l+m}{2} ../ref/Styphi\PYZus{}CT18.fa \PY{l+s+se}{\PYZbs{}}
../fastq/ERR5243693\PYZus{}1.trim.fastq.gz ../fastq/ERR5243693\PYZus{}2.trim.fastq.gz \PY{l+s+se}{\PYZbs{}}
\PYZgt{} ERR5243693.sam
\end{Verbatim}
\end{tcolorbox}

    This may take a few minutes to run. Let's take a look at the options:

\begin{itemize}
\tightlist
\item
  \texttt{-t} : tells the program to use 2 CPUs
\item
  \texttt{../ref/Styphi\_CT18.fa} is the reference to match the reads
  against
\item
  \texttt{../fastq/ERR5243693\_1.trim.fastq.gz\ and\ ../fastq/ERR5243693\_2.trim.fastq.gz}
  are the fastq files containing our sequence reads (after trimming) for
  sample ERR5243693
\item
  the output is SAM format and \texttt{\textgreater{}\ ERR5243693.sam}
  redirects the output to the file \texttt{ERR5243693.sam}
\end{itemize}

When complete, convert the sam file to a bam file with samtools:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools view \PYZhy{}@ \PY{l+m}{2} \PYZhy{}bh \PYZhy{}o ERR5243693.bam ERR5243693.sam
\end{Verbatim}
\end{tcolorbox}

    The \texttt{-@} option tells the program to use 2 CPUs and the
\texttt{-b} option specifies to write the output as a bam file,
\texttt{-h} option means include the header information in the output.
The \texttt{-o} option specifies the name of the output file
\texttt{ERR5243693.bam}.

Sort the bam file and then index the sorted bam file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools sort \PYZhy{}@ \PY{l+m}{1} \PYZhy{}o ERR5243693.sorted.bam \PYZhy{}T ERR5243693.sorted ERR5243693.bam
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools index ERR5243693.sorted.bam
\end{Verbatim}
\end{tcolorbox}

    Generate some statistics about the alignment:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools stats ERR5243693.sorted.bam \PYZgt{} ERR5243693.stats
samtools flagstat ERR5243693.sorted.bam \PYZgt{} ERR5243693.flagstat
samtools coverage ERR5243693.sorted.bam \PYZgt{} ERR5243693.coverage
\end{Verbatim}
\end{tcolorbox}

    Now repeat for the other samples.

\hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  What \%reads mapped to the reference for each sample?
\item
  What \%genome was covered for each sample?
\item
  What is the mean depth of coverage for each sample?
\end{enumerate}

    \hypertarget{call-variants}{%
\subsubsection{Call variants}\label{call-variants}}

Go through each position in the reference genome and look at reads
aligned at that position and make a call about what the base is at that
position for the sample. This information will be stored in a VCF file
and if there are any differences then this will be marked as a variant
(snp/indel) in this VCF file.

Again create a directory to store the results of the variant calling
step:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
mkdir ../variants
\PY{n+nb}{cd} ../variants
\end{Verbatim}
\end{tcolorbox}

    Run this for sample ERR5243693 using bcftools:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools mpileup \PYZhy{}\PYZhy{}fasta\PYZhy{}ref ../ref/Styphi\PYZus{}CT18.fa \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}min\PYZhy{}BQ \PY{l+m}{20} \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}annotate \PY{l+s+se}{\PYZbs{}}
FORMAT/AD,FORMAT/ADF,FORMAT/ADR,FORMAT/DP,FORMAT/SP,INFO/AD,\PY{l+s+se}{\PYZbs{}}
INFO/ADF,INFO/ADR ../samtools/ERR5243693.sorted.bam \PY{p}{|} bcftools call \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}output\PYZhy{}type v \PYZhy{}\PYZhy{}ploidy \PY{l+m}{1} \PYZhy{}\PYZhy{}multiallelic\PYZhy{}caller \PYZhy{} \PY{p}{|} \PY{l+s+se}{\PYZbs{}}
bcftools view \PYZhy{}\PYZhy{}output\PYZhy{}file ERR5243693.vcf.gz \PYZhy{}\PYZhy{}output\PYZhy{}type z
\end{Verbatim}
\end{tcolorbox}

    Again this may take some time to run. Let's look at the options:

The \texttt{bcftools\ mpileup} command is passed the following options:

\begin{itemize}
\tightlist
\item
  \texttt{-\/-fasta-ref\ ../ref/Styphi\_CT18.fa} is the reference that
  the reads were mapped to
\item
  \texttt{-\/-min-BQ\ 20} is minimum base quality for a base to be
  considered
\item
  \texttt{-\/-annotate\ FORMAT/AD,FORMAT/ADF,FORMAT/ADR,FORMAT/DP,FORMAT/SP,INFO/AD,INFO/ADF,INFO/ADR}
  tells the program what tags to incude in the VCF
\end{itemize}

The VCF produced by the bcftools mpileup command is passed to
\texttt{bcftools\ call} with the following options:

\begin{itemize}
\tightlist
\item
  \texttt{-\/-output-type\ v} tells the program to output VCF format
\item
  \texttt{-\/-ploidy\ 1} tells the program that we are daling with
  haploid datasets
\item
  \texttt{-\/-multiallelic-caller} tells the program which variant
  calling algorithm to use
\end{itemize}

The VCF produced by the \texttt{bcftools\ call} command is passed to
\texttt{bcftools\ view} with the following options which converts the
VCF to a compressed VCF file:

\begin{itemize}
\tightlist
\item
  \texttt{-\/-output-file\ ERR5243693.vcf.gz} tells the program the name
  of the output file
\item
  \texttt{-\/-output-type\ z} tells the program to create a gzipped VCF
  file as output
\end{itemize}

When it completes, index the gzipped VCF file that has been created:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
tabix \PYZhy{}p vcf \PYZhy{}f ERR5243693.vcf.gz
\end{Verbatim}
\end{tcolorbox}

    Take a look at the VCF file that was produced:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools view ERR5243693.vcf.gz \PY{p}{|} less \PYZhy{}S
\end{Verbatim}
\end{tcolorbox}

    Notice that the VCF has information about every site in the genome.

    Now generate some statistics about the VCF file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools stats ERR5243693.vcf.gz \PYZgt{} ERR5243693.vcf.stats.txt
\end{Verbatim}
\end{tcolorbox}

    Now repeat for the other samples.

Look at the statistics for the variant calling:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
less ERR5243693.vcf.stats.txt
less ERR5243695.vcf.stats.txt
less ERR5243699.vcf.stats.txt
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many sites are in the VCF file for each sample?
\item
  Does this match to the size of the reference used in the read mappping
  step?
\item
  How many variant sites were identified for each sample?
\end{enumerate}

    \hypertarget{filter-the-vcfs}{%
\subsubsection{Filter the VCFs}\label{filter-the-vcfs}}

We want to identify calls where we have a high confidence that they are
correct (and not due to sequencing errors and/or misalignment of the
reads). We use criteria like read depth at a position, quality scores
etc. to filter out low quality calls at each position.

Use bcftools to filter sites for sample ERR5243693.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools filter \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}output ERR5243693.filtered.vcf.gz \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}soft\PYZhy{}filter LowQual \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}exclude \PY{l+s+s2}{\PYZdq{}QUAL\PYZlt{}25 || FORMAT/DP\PYZlt{}10 || MAX(FORMAT/ADF)\PYZlt{}2 || MAX(FORMAT/ADR)\PYZlt{}2 \PYZbs{}}
\PY{l+s+s2}{|| MAX(FORMAT/AD)/SUM(FORMAT/DP)\PYZlt{}0.9 || MQ\PYZlt{}30 || MQ0F\PYZgt{}0.1\PYZdq{}} \PY{l+s+se}{\PYZbs{}}
\PYZhy{}\PYZhy{}output\PYZhy{}type z ERR5243693.vcf.gz
\end{Verbatim}
\end{tcolorbox}

    The filtering step may take some time to run. Let's look at the options
used in the command above:

\begin{itemize}
\tightlist
\item
  \texttt{-\/-output} specifies the name of the output file
\item
  \texttt{-\/-soft-filter} tells the program to keep any filtered
  position in the file and mark them as \texttt{LowQual} rather than
  remove them completely from the file (hard filter)
\item
  \texttt{-\/-exclude} lists the filtering criteria to apply to each
  position
\item
  \texttt{-\/-output-type} specifies the type of output file to create,
  in this case z means a xompressed VCF file
\end{itemize}

When the bcftools filter command is complete, index the filtered VCF
file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
tabix \PYZhy{}p vcf \PYZhy{}f ERR5243693.filtered.vcf.gz
\end{Verbatim}
\end{tcolorbox}

    Take a look at the VCF file and notice how some of the sites are marked
as \texttt{PASS} or \texttt{LowQual} under the filter column.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools view ERR5243693.filtered.vcf.gz \PY{p}{|} less \PYZhy{}S
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

Looking at the filtered VCF for sample ERR5243693:\\
1. Does position 2 pass or fail?\\
2. Does position 244 pass or fail?\\
3. What is the reason for the failure at position 2527?

    To view only the sites that passed the filtering step:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools view \PYZhy{}f PASS ERR5243693.filtered.vcf.gz \PY{p}{|} less
\end{Verbatim}
\end{tcolorbox}

    To generate some statistics about the filtered VCF file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bcftools stats ERR5243693.filtered.vcf.gz \PYZgt{} \PY{l+s+se}{\PYZbs{}}
ERR5243693.filtered.stats.txt
\end{Verbatim}
\end{tcolorbox}

    Now repeat for the other samples.

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\tightlist
\item
  How many sites were marked as low quality in the filtering step?
\item
  How many variant sites were marked as low quality in the filtering
  step?
\end{enumerate}

    \hypertarget{call-a-consensus-sequence-for-each-sample}{%
\subsubsection{Call a consensus sequence for each
sample}\label{call-a-consensus-sequence-for-each-sample}}

A pseudogenome is a reconstruction of what we think the genome is for
the sample using the reference genome as a basis. To create it for a
sample, you go through each position in the reference and determine what
base is called (using the VCF from the previous steps) for the sample.
Sometimes this will be the same as the reference, and sometimes it will
differ from the reference (a variant). For positions that are flagged as
low quality/filtered out (e.g.~no reads covering the position) we use an
N in the pseudogenome. This is because you cannot be confident what the
base is at this position for the sample. In the end the length of the
pseudogenome for your sample should be the same as the length of the
reference.

To create a pseudogenome for sample ERR5243693 use the script
\textit{vcf2pseudogenome.pl}. This has already been installed on the
computer using this command (it also has a dependency on \textit{pysam}
and \textit{biopython}):

\texttt{wget\ https://raw.githubusercontent.com/nf-core/bactmap/master/bin/vcf2pseudogenome.py}

Create a directory to store the pseudogenomes:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
mkdir ../pseudogenomes
\PY{n+nb}{cd} ../pseudogenomes
\end{Verbatim}
\end{tcolorbox}

    The run the \texttt{vcf2pseudogenome.py} script:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
vcf2pseudogenome.py \PYZhy{}r ../ref/Styphi\PYZus{}CT18.fa \PY{l+s+se}{\PYZbs{}}
\PYZhy{}b ../variants/ERR5243693.filtered.vcf.gz \PYZhy{}o ERR5243693.fas
\end{Verbatim}
\end{tcolorbox}

    Now repeat for the other samples.

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  What is the length of the pseudogenomes? (Hint: Use assembly-stats)
\item
  Does it match the length of the reference?
\end{enumerate}

    \hypertarget{create-a-multiple-sequence-alignment-of-all-pseudogenomes}{%
\subsubsection{Create a multiple sequence alignment of all
pseudogenomes}\label{create-a-multiple-sequence-alignment-of-all-pseudogenomes}}

Remember that to reconstruct the phylogeny of our samples we need a
multifasta alignment of our sequences. We are going to use a
reconstruction of the entire genome as our set of sequences.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat *.fas \PYZgt{} aligned\PYZus{}pseudogenomes.aln
\end{Verbatim}
\end{tcolorbox}

    Because of the way the pseudogenomes were constructed resulting in them
all being the same length we do not have to perform a multi-sequence
alignment to ensure all the bases are homologous.

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many sequences in the multiple sequence alignment file of
  pseudogenomes?
\item
  What is the largest and mean sequence length? (Hint: Use
  assembly-stats)
\end{enumerate}

    \hypertarget{add-the-reference-genome-to-the-multiple-sequence-alignment}{%
\subsubsection{Add the reference genome to the multiple sequence
alignment}\label{add-the-reference-genome-to-the-multiple-sequence-alignment}}

Sometimes you may want to include the reference genome in your tree. To
do this, add the reference sequence to the multifasta alignment of your
samples.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat ../ref/Styphi\PYZus{}CT18.fa aligned\PYZus{}pseudogenomes.aln \PY{l+s+se}{\PYZbs{}}
\PYZgt{} ref\PYZus{}and\PYZus{}aligned\PYZus{}pseudogenomes.aln
\end{Verbatim}
\end{tcolorbox}

    At this point it would be useful to look at your alignment in a multiple
sequence alignment viewer e.g.~seaview.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
seaview ref\PYZus{}and\PYZus{}aligned\PYZus{}pseudogenomes.aln \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many sequences in the multiple sequence alignment file containng
  the reference genome and sample pseudogenomes?
\end{enumerate}

    \hypertarget{mask-repetitive-regions-optional-but-good-practice}{%
\subsubsection{Mask repetitive regions (optional but good
practice)}\label{mask-repetitive-regions-optional-but-good-practice}}

Bases called in repetitive regions may not be true variation (e.g.~due
to misalignment of reads) and may compromise the core phylogenetic
signal. Therefore it is good practice to identify known repetitive
regions and mask these out from your alignment.

To achieve this, either a file of known regions for the reference you
mapped to will exist (see the literature) or one can be generated by
matching the reference genome against itself (to identify repeat
regions) with e.g.~Mummer and Phast used to identify prophage. You can
use the software remove\_blocks to mask out any repetitive region in
your multifasta alignment, this masking involves replacing the bases
with Ns in this region. To save time we wont run this step here.

    \hypertarget{draw-a-tree-with-iqtree}{%
\subsubsection{Draw a tree with iqtree}\label{draw-a-tree-with-iqtree}}

Now that we have a multiple sequence alignment, we can use IQ-TREE to
build a maximum likelihood phylogeny. But first create a directory to
store the iqtree analysis.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
mkdir ../iqtree
\PY{n+nb}{cd} ../iqtree
\end{Verbatim}
\end{tcolorbox}

    Calculating a phylogeny on whole genome sequences can be very time
consuming. We can speed this up by only using the variable sites (SNPs).
These are sites that differ in at least one of the samples. However, we
need to be aware that only including variable sites can affect the
evolutionary rate estimates made by phylogenetics software - therefore,
we need to account for the sites we remove in our analysis.

We will use snp-sites to do this. You can view the options for
snp-sites:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
snp\PYZhy{}sites \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    First, remove all the invariant sites and create a SNP-only multiple
sequence alignment:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
snp\PYZhy{}sites \PYZhy{}o pseudogenomes.snps.aln \PY{l+s+se}{\PYZbs{}}
../pseudogenomes/ref\PYZus{}and\PYZus{}aligned\PYZus{}pseudogenomes.aln
\end{Verbatim}
\end{tcolorbox}

    Look at the new file that is created:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
less pseudogenomes.snps.aln
\end{Verbatim}
\end{tcolorbox}

    We can also see how many invariant sites were removed (and what
proportion of A, T, G, C they were) using

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
snp\PYZhy{}sites \PYZhy{}C ../pseudogenomes/ref\PYZus{}and\PYZus{}aligned\PYZus{}pseudogenomes.aln
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{check-your-understanding}{%
\paragraph{Check your understanding}\label{check-your-understanding}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many variant sites are identified?
\item
  How many invariant sites are identified?
\item
  Does this correlate with the expected number i.e.~from the literature?
\end{enumerate}

Now look at the options for IQ-TREE:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
iqtree \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    And then construct the tree with IQ-TREE:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
iqtree \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}s pseudogenomes.snps.aln \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}fconst \PY{k}{\PYZdl{}(}snp\PYZhy{}sites \PYZhy{}C ../pseudogenomes/ref\PYZus{}and\PYZus{}aligned\PYZus{}pseudogenomes.aln\PY{k}{)} \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}m GTR+G \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}B \PY{l+m}{1000} \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}T \PY{l+m}{2} \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}mem 2GB
\end{Verbatim}
\end{tcolorbox}

    In the command above, we have used te following options

\begin{itemize}
\tightlist
\item
  \texttt{-s} to specify the multiple sequence alignment file
  \texttt{pseudogenomes.snpsites.aln}
\item
  \texttt{-fconst} to ask IQ-TREE to take account of missing invariant
  sites \$(snp-sites -C ref\_and\_aligned\_pseudogenomes.aln calculates
  the specific values to pass to iqtree)
\item
  \texttt{-m} to specify an evolutionary model, we want IQ-TREE to use
  -m GTR+G
\item
  \texttt{-B} to perform 1000 ultrafast bootstraps
\item
  \texttt{-T} to tell IQ-TREE to use a maximum of 2 CPUs (threads)
\item
  \texttt{-mem} 2GB to tell IQ-TREE to use a maximum of of 2GB memory
\end{itemize}

    Our maximum likelihood tree is labelled
\texttt{pseudogenomes.snpsites.aln.treefile}. The treefile suffix is not
always correctly identified by many tools, so we'll relabel this as
something else:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cp pseudogenomes.snps.aln.treefile pseudogenomes.snps.aln.tre
\end{Verbatim}
\end{tcolorbox}

    We can look at the raw text file:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat pseudogenomes.snps.aln.tre
\end{Verbatim}
\end{tcolorbox}

    But it is better if we visualise this using a tree view e.g.figtree

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
figtree pseudogenomes.snps.aln.tre \PY{p}{\PYZam{}}
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{accounting-for-recombination-with-gubbins}{%
\subsubsection{Accounting for recombination with
Gubbins}\label{accounting-for-recombination-with-gubbins}}

Variation due to recombination events can mask the core phylogenetic
signal, therefore it is recommended to identify these regions in your
alignment and mask them out. We can use gubbins to infer recombining
sites by looking for increased SNP density that occurs in specific
ancestral nodes. Look at the options for gubbins:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
run\PYZus{}gubbins.py \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    Make a directory to store the results of running gubbins:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
mkdir ../gubbins
\PY{n+nb}{cd} ../gubbins
\end{Verbatim}
\end{tcolorbox}

    Run gubbins on your data:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
run\PYZus{}gubbins.py \PYZhy{}c \PY{l+m}{2} \PYZhy{}p gubbins \PY{l+s+se}{\PYZbs{}}
../pseudogenomes/ref\PYZus{}and\PYZus{}aligned\PYZus{}pseudogenomes.aln
\end{Verbatim}
\end{tcolorbox}

    The \texttt{-c} option tells the program to use 2 CPUs and the
\texttt{-p} option tells the program to name all files with the prefix
gubbins. This command can take a few minutes to run.

\textbf{Note:} Gubbins must be run on the entire alignment (not just
SNPs) as it uses the spatial distribution to identify and filter out
recombinant regions.

If gubbins takes more than 10 mins to complete, we have already run it
for you - the files are available at:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls \PYZhy{}l \PYZti{}/course\PYZus{}data/snp\PYZhy{}phylogeny/data/typhi/gubbins\PYZus{}backup
\end{Verbatim}
\end{tcolorbox}

    Lets look at what gubbins has done:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls \PYZhy{}l *
\end{Verbatim}
\end{tcolorbox}

    You can explore these files. For example
\texttt{gubbins.recombination\_predictions.gff} is a gff file that
contains a record of each recombination block identified, how many SNPs
it contains, and what samples are affected.

    head gubbins.recombination\_predictions.gff

    The file \texttt{gubbins.final\_tree.tre} is a phylogeny in which
recombination has already been accounted for.

You can visualise this in figtree.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
figtree gubbins.final\PYZus{}tree.tre
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{root-the-phylogeny}{%
\subsection{Root the phylogeny}\label{root-the-phylogeny}}

The trees generated by iqtree and gubbins are unrooted, but we may want
to apply some evolutionary direction to them. One strategy for rooting a
tree is called \textit{midpoint rooting}. Midpoint rooting involves
locating the midpoint of the longest path between any two tips and
putting the root in that location. Note that this does not necessarily
infer the true root, and this should be used with caution.

To midpoint root our tree, we will use a simple script written in python
that uses the ete package. You can examine the code:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
less ../midpoint.root.py
\end{Verbatim}
\end{tcolorbox}

    \begin{figure}
\centering
\includegraphics{img/midpoint.root.image.png}
\caption{Python script to midpoint root a tree}
\end{figure}

    Run this script to midpoint root the tree.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
python midpoint.root.py gubbins.final\PYZus{}tree.tre \PYZgt{} gubbins.final\PYZus{}tree.midpoint.tre
\end{Verbatim}
\end{tcolorbox}

    Visualise this in figtree.

Another common strategy for rooting the tree is \textit{outgroup rooting}.
This is the preferred approach for bacterial datasets. Outgroup rooting
involves including one or more sequences in the analysis that are more
distantly related to our sequences of interest than they are to one
another. These sequence are usually referred to as \textit{outgroups}. The
root estimate is then simply the point at which the outgroup(s) join the
tree. The best possible outgroups are those available which are most
closely related to our sequences of interest but still different enough.
For examples, in this dataset we could use a \textit{Salmonella paratyhi A
sample} as an outgroup.

    There are a few ways to do this. One is to obtain sequence data for the
outgroup sample and incorporate it into the dataset from the beginning
of the analysis and construct a pseudogenome for the outgroup. If no
sequence data is available, you can take a complete reference genome for
the outgroup and \texttt{shred} it to simulate sequencing reads for the
reference. Just remember when calculating and reporting the number of
variant sites for your dataset that you remove the outgroup from the
alignment.

    \hypertarget{clean-up}{%
\subsubsection{Clean-up!}\label{clean-up}}

Clean up any intermediate files that were generated during the analysis
that you no longer require. This is always an important last step of any
analysis as sequence data analysis files can use up large amounts of
disk space. Let's look at what files were created in our analysis

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ..
ls \PYZhy{}alhrt */*
\end{Verbatim}
\end{tcolorbox}

    As you can see the data analysis has generated a lot of files! So let's
remove any files that do not need to be kept long term:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
rm fastq/*.trim.fastq.gz
rm fastq/*.fastp.html
rm fastq/*.fastp.json
rm samtools/*.sam
rm samtool/ERR5243693.bam*
rm samtools/ERR5243695.bam*
rm samtools/ERR5243699.bam*
rm variants/ERR5243693.vcf.gz*
rm variant/ERR5243695.vcf.gz*
rm variants/ERR5243699.vcf.gz*
\end{Verbatim}
\end{tcolorbox}

    Now go to the next section: \href{metadata.ipynb}{Phylogeny and
Metadata}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{phylogeny-and-metadata}{%
\section{Phylogeny and Metadata}\label{phylogeny-and-metadata}}

Often it is useful to visualise metadata about your samples in the
context of your phylogenetic tree. For example, location information can
provide insight into the processes that drive their epidemiology. This
can be used to infer whether single introductions of a pathogen have
occurred followed by local evolution or whether it transmits frequently
across borders. It can also indicate regions affected by antimicrobial
resistance (AMR).

In this section, you will use \texttt{Microreact} to visualise the
phylogeny you created in the previous section with some basic metadata.
\texttt{Microreact} is software developed by the Centre for Genomic
Pathogen Surveillance (CGPS) that allows you to upload, visualise and
explore any combination of clustering (trees), geographic (map) and
temporal (timeline) data. Other metadata variables are displayed in a
table. You can specify colours and/or shapes to display on the map, tree
and/or timeline. A permanent URL is produced for you to share your
Microreact, or a .microreact file can be downloaded for sharing with
collaborators.

There are several alternatives including FigTree, Phandango and iToL but
for the purposes of this tutorial we will demostrate the use of
Microreact for metadata visualisation.

    \hypertarget{the-dataset}{%
\subsection{The dataset}\label{the-dataset}}

In this section we will make use of a maximum likelihood phylogenetic
tree of the \textit{S. typhi} isolates from the previous section and a
metadata file with information on location, time of isolation and
genotype for the samples.

    \hypertarget{create-a-microreact-project}{%
\subsection{Create a Microreact
project}\label{create-a-microreact-project}}

Start by opening up a new window in Firefox and typing
https://Microreact.org in the address bar. Click on ``Upload'' as shown
below and create a new project.


\begin{center}
\includegraphics[Microreact front page]{img/microreact-front-page.png}
\end{center}


Open a file browser and navigate to the metadata directory
(\texttt{/home/manager/course\_data/snp-phylogeny/data/metadata}). Drag
and drop the files ``metadata.csv'' and ``tree.tre'' from the file
browser onto your Internet browser.


\begin{center}
\includegraphics[Uploading files to Microreact]{img/microreact-upload.png}
\end{center}


Once the tree and metadata files are loaded you will be directed to a
new window where files will be automatically detected as Data (CSV or
TSV) file (metadata.csv) and Tree (Newick) file (tree.tre). In this new
window click on `Continue'. In the next window and make sure column `ID'
is selected as the `ID column' and then click on `Continue'.


\begin{center}
\includegraphics[Selecting sample identifier]{img/microreact-id.png}
\end{center}


Once these forms are completed your data will be utilized to create a
Microreact project. You should now have a view similar to the ones shown
below. You should see a tree view and a timeline and metadata view.


\begin{center}
\includegraphics[Microreact timeline view]{img/microreact-timeline.png}
\end{center}



\begin{center}
\includegraphics[Microreact metadata view]{img/microreact-metadata.png}
\end{center}


    \hypertarget{exploring-the-metadata}{%
\subsection{Exploring the metadata}\label{exploring-the-metadata}}

Clicking on `Show Controls' button in the top right hand corner of the
the tree window allows you to view different kinds of trees and change
text/node size. Try changing layout of the tree.


\begin{center}
\includegraphics[Uploading files to Microreact]{img/microreact-options.png}
\end{center}


There are lots more things you can do with \texttt{Microreact} but
unfortunately we don't have time to cover them in this tutotial.

\textbf{Congratulations}, you have reached the end of the tutoial!


    % Add a bibliography block to the postdoc



\end{document}
